### ISSUE TITLE
Harmonize nodata usage

### ISSUE THREAD

**Issue by johanvdw** [inbo/niche_vlaanderen#375]
There are too many different nodata definitions in use in niche:
- if data is float 32: np.nan is use
- if data is uint8: 255 is used
- in other cases -99 is used

I think we should consider moving all data types to float32. Areas are relatively small, and we compress tif files anyway. This would also mean we can use np.nan for no data, which propagates properly without special tweaks.

[masked arrays](https://numpy.org/doc/stable/reference/maskedarray.html) are another option to remove this low level code, but fixing the data type is probably even better.

Note that currently the code seems to be working well, but it is rather complex internally, leading eg to #335 .

**Comment by cecileherr** [comment]
A side note: in the past (with niche 1.2, Win 10, 8Go RAM but hardly any free hard disk < 10 Go) I have had problems with memory issues with some projects (example: for a project with resolution of 5*5 m 

> MemoryError: unable to allocate 43.8 MiB for an array with shape (2760, 4160) and data type float32

).

I suspect changing the data type to float/solving issue #335 might lead to more memory problems (?). Would it be possible to test this/give an idea of the impact on memory/speed? Thx!

**Comment by stijnvanhoey** [comment]
@johanvdw  can you provide some input on:

-  the https://github.com/inbo/niche_vlaanderen/blob/master/niche_vlaanderen/codetables.py#L220-L222 addition of both -99 and 255 to the allowed values? How should this be translated when converting to float32 for all? On which level does this function operate, input from a user or rather on an package internal management level?
- how does https://github.com/inbo/niche_vlaanderen/blob/master/niche_vlaanderen/acidity.py#L15 relate to https://github.com/inbo/niche_vlaanderen/blob/master/niche_vlaanderen/acidity.py#L70, https://github.com/inbo/niche_vlaanderen/blob/master/niche_vlaanderen/acidity.py#L82,... using both 255, 99 ad -99 for the same data type?

**Comment by stijnvanhoey** [comment]
I did a check on the usage of masked arrays, see https://github.com/inbo/niche_vlaanderen/pull/387, but I will reverse the masked array implementation. Whereas the usage simplifies the implementation an no-data-regions considerably, the reasons are:

- https://numpy.org/doc/stable/reference/module_structure.html#legacy-namespaces mentions numpy.ma as deprecated module (note, this is not mentioned in https://numpy.org/doc/stable/reference/maskedarray.html) which requires overhaul
- reports of others that masked arrays are slower (e.g. https://github.com/Unidata/netcdf4-python/issues/809). I did test this out myself on the current code and the decrease in speed is considerable, e.g.

    - unit test "test_zwarte_beek" -> 2.9s (median) voor current master (no masked arrays) versus 7.7s (median) voor masked-array approach
    - single niche modle run in getting-started notebook -> 623 ms ± 118 ms voor current master versus 1.93 s ± 506 ms voor masked-array approach

As this is more than double the time, we do no longer consider this a valuable option and focus on predefined data types (uint8 and float32) for each variable with clear no-data implementation (255 and np.nan for uint8 and float32 respectively)
